functions/go/create-setters/createsetters/create_setters.go

```diff
@@ -0,0 +1,275 @@
+
+environments: # kpt-set: ${env}
+  - dev
+  - stage
+
+e.g. for input of Mapping node with FlowStyle
+
+env: [foo, bar]
+
+For input ApplySetters [Name: env, ValueSet: [foo, bar]], yaml node is transformed to
+
+env: [foo, bar] # kpt-set: ${env}
+*/
+
+func (cs *CreateSetters) visitMapping(object *yaml.RNode, path string) error {
+	return object.VisitFields(func(node *yaml.MapNode) error {
+		if node == nil || node.Key.IsNil() || node.Value.IsNil() {
+			// don't do IsNilOrEmpty check cs empty sequences are allowed
+			return nil
+		}
+		
+		// the aim of this method is to create-setter for sequence nodes
+		if node.Value.YNode().Kind != yaml.SequenceNode {
+			// return if it is not a sequence node
+			return nil
+		}
+
+		// checks if the kind is flowstyle and adds comment to its value node 
+		// else it adds the comment to the key node 
+		changeNodeComment := node.Key
+
+		if node.Value.YNode().Style == yaml.FlowStyle {
+			changeNodeComment = node.Value
+		}
+
+		// add the key to the field path
+		fieldPath := strings.TrimPrefix(fmt.Sprintf("%s.%s", path, node.Key.YNode().Value), ".")
+
+		elements, err := node.Value.Elements()
+		if err != nil {
+			return errors.Wrap(err)
+		}
+		// extracts the values in sequence node to an array 
+		var nodeValues []string
+		for _, values := range elements {
+			nodeValues = append(nodeValues, values.YNode().Value)
+		}
+
+		for _, arraySetters := range cs.ArraySetters {
+			// checks if all the values in node are present in array setter
+			if checkEqual(nodeValues, arraySetters.ValueSet) {
+				changeNodeComment.YNode().LineComment = fmt.Sprintf("kpt-set: ${%s}", arraySetters.Name)
+				return nil
+			}
+		}
+
+		cs.Results = append(cs.Results, &Result{
+			FilePath:  cs.filePath,
+			FieldPath: fieldPath,
+			Value: changeNodeComment.YNode().LineComment,
+		})
+		return nil
+	})
+}
+
+/**
+visitScalar accepts the input scalar node and performs following steps,
+checks if it is a scalar node 
+checks if the path ends with ']' (which is the value of a sequence node)
+adds the linecomment if it's value matches with any of the setter 
+
+e.g.for input of scalar node 'nginx:1.7.1' in the yaml node
+
+apiVersion: v1
+...
+  image: nginx:1.7.1
+
+and for input CreateSetters [[name: image, value: nginx], [name: tag, value: 1.7.1]]
+The yaml node is transformed to
+
+apiVersion: v1
+...
+  image: nginx:1.7.1 # kpt-set: ${image}:${tag}
+
+*/
+
+func (cs *CreateSetters) visitScalar(object *yaml.RNode, path string) error {
+	if object.YNode().Kind != yaml.ScalarNode {
+		// return if it is not a scalar node
+		return nil
+	}
+
+	// checks if the path ends with "]"
+	if path[len(path)-1] == ']' {
+		return nil
+	}
+
+	// its a flag to indicate if value matches with any of the setter 
+	contains := false
+
+	linecomment := object.YNode().Value
+
+	for _, setter := range cs.Setters {
+
+		if strings.Contains(object.YNode().Value, setter.Value) {
```

Comment:
There are some corner cases with this approach.

What if the input is 
1. [[name: image, value: nginx], [name: foo, value: image]]
2. [[name: image, value: nginx], [name: foo, value: nginx-abc]]

Think of other cases it could fail.

Reply:
There is not a right way to solve 2 for now. We can solve that in the future based on customer feedback.

You can solve for 1 though. Don't replace the value if it is within ${}

And separate this logic into a method which takes cs.Setters, object.YNode().Value and returns the setter comment. Add unit tests to that method.

Reply:
Sure